stats.go

```diff
@@ -450,74 +484,169 @@ func (s *statStore) NewPerInstanceTimer(name string, tags map[string]string) Tim
+	return s.registry.newTimerWithTagSet(joinScopes(s.name, name), s.mergeTags(tags))
 }
 
-func (s subScope) NewPerInstanceTimer(name string, tags map[string]string) Timer {
-	return s.registry.NewPerInstanceTimer(joinScopes(s.name, name), s.mergeTags(tags))
+func (s *subScope) NewPerInstanceTimer(name string, tags map[string]string) Timer {
+	return s.registry.newTimerWithTagSet(joinScopes(s.name, name),
+		s.mergePerInstanceTags(tags))
 }
 
 func joinScopes(parent, child string) string {
 	return parent + "." + child
 }
 
-// mergeTags augments tags with all scope-level tags that are not already present.
-// Modifies and returns tags directly.
-func (s subScope) mergeTags(tags map[string]string) map[string]string {
-	if len(s.tags) == 0 {
-		return tags
+// mergeOneTag is an optimized for inserting 1 tag and will panic otherwise.
+func (s *subScope) mergeOneTag(tags map[string]string) tagSet {
+	if len(tags) != 1 {
+		panic("invalid usage")
 	}
-	if len(tags) == 0 {
+	var p tagPair
+	for k, v := range tags {
+		p = tagPair{key: k, value: replaceChars(v)}
+		break
+	}
+	if p.key == "" || p.value == "" {
 		return s.tags
 	}
-	for k, v := range s.tags {
-		if _, ok := tags[k]; !ok {
-			tags[k] = v
+	a := make(tagSet, len(s.tags), len(s.tags)+1)
+	copy(a, s.tags)
+	return a.Insert(p)
+}
+
+// mergeTags returns a tagSet that is the union of subScope's tags and the
+// provided tags map. If any keys overlap the values from the provided map
+// are used.
+func (s *subScope) mergeTags(tags map[string]string) tagSet {
+	switch len(tags) {
+	case 0:
+		return s.tags
+	case 1:
+		// optimize for the common case of there only being one tag
+		return s.mergeOneTag(tags)
+	default:
+		// write tags to the end of the scratch slice
+		scratch := make(tagSet, len(s.tags)+len(tags))
+		a := scratch[len(s.tags):]
+		i := 0
+		for k, v := range tags {
+			if k != "" && v != "" {
+				a[i] = tagPair{key: k, value: replaceChars(v)}
+				i++
+			}
+		}
+		a = a[:i]
+		a.Sort()
+
+		if len(s.tags) == 0 {
+			return a
 		}
+		return mergeTagSets(s.tags, a, scratch)
+	}
+}
+
+// mergePerInstanceTags returns a tagSet that is the union of subScope's
+// tags and the provided tags map with. If any keys overlap the values from
+// the provided map are used.
+//
+// The returned tagSet will have a per-instance key ("_f") and if neither the
+// subScope or tags have this key it's value will be the default per-instance
+// value ("i").
+//
+// The method does not optimize for the case where there is only one tag
+// because it is used less frequently.
+func (s *subScope) mergePerInstanceTags(tags map[string]string) tagSet {
+	if len(tags) == 0 {
+		if s.tags.Contains("_f") {
+			return s.tags
+		}
+		// create copy with the per-instance tag
+		a := make(tagSet, len(s.tags), len(s.tags)+1)
+		copy(a, s.tags)
+		return a.Insert(tagPair{key: "_f", value: "i"})
+	}
+
+	// write tags to the end of scratch slice
+	scratch := make(tagSet, len(s.tags)+len(tags)+1)
```

Comment:
Why use a scratch slice instead of `a := make(tagSet, len(s.tags)+len(tags)+1)`?

Reply:
```go
// mergeTagSets merges s1 into s2 and stores the result in scratch. Both s1 and
// s2 must be sorted and s2 can be a sub-slice of scratch if it is located at
// the tail of the slice (this allows us to allocate only one slice).
func mergeTagSets(s1, s2, scratch tagSet) tagSet {
	...
}
```
We're using `scratch` and making `a` a subset of it (at the tail of the slice) so that we can save an allocation. If we put `a` at the head of `scratch` then we'd overwrite values during the merge. Below is what the entire merge process looks like:

```
s.pairs:     [1, 3, 5]
tags (keys): [2, 4, 6]
scratch:     [0, 0, 0, 0, 0, 0] # the tags are written to the end of the
a:                    [0, 0, 0] # sub-slice of scratch
# write tags to `a` and sort
a:                    [2, 4, 6]
scratch:     [0, 0, 0, 2, 4, 6]
# now when we merge `s.pairs` into `a` we are writing to the head of `scratch`
# first iteration of the merge loop
scratch:     [1, 0, 0, 2, 4, 6]
# second iteration of the merge loop
scratch:     [1, 2, 0, 2, 4, 6]
# after all iterations of the merge loop
scratch:     [1, 2, 3, 4, 5, 6]
```